### DISCUSSION TITLE
Regress an expression of control laws

### DISCUSSION THREAD

**Discussion by JazzyFeng** [MilesCranmer/PySR#881] [Q&A]
First thanks for your excellent open-sourced work. I am using pySR to find the underneath expression in control problem. Currently, I use a Reinforcement Learning algorithm DDPG to train a neural network to stabilize  Pendulum and keep it vertical. Now, I randomly generated states (Pendulum x, y in cartesian coordinates and angular velocity) and feed forward them to DDPG and got output, namely, actions. I use pySR to fit states and actions, but the result show that loss is high and score is low. 
Binary operator: + - * /
Unary operator: cos sine
I also use plot to visualize actions Truth and predictions by pySR, and it give me a plot below.  Can you give me some advice on how to improve the regression result?
![output](https://github.com/user-attachments/assets/d87cad6e-94e6-40df-a3f6-41c61cea7408)

**Comment by MilesCranmer** [comment]
Have a look through the tuning page here: https://ai.damtp.cam.ac.uk/pysr/tuning/

**Reply by JazzyFeng** [reply]
Thank you for your prompt reply！I checked the tuning page and do it step by step. 
Another question: Since I tried to load the trained torch model, I use notebook in Pycharm. So I cannot early stop training in iPython by press 'q' and 'enter'.

**Reply by MilesCranmer** [reply]
You can set up early stopping with timeout or max evals too: https://ai.damtp.cam.ac.uk/pysr/api/#stopping-criteria

**Reply by JazzyFeng** [reply]
Your suggestion works! 
Just let you know my progression. Now, I use the output formula to control the pendulum, and it works occasionally. I think the jittery curve suggests that prediction is overestimating or underestimating sometimes. If I set a loss threshold in `early_stop_condition` and increase `niterations` , will it converge to a very low loss?

**Reply by JazzyFeng** [reply]
Thank you for your work. I can access the best equation by get_best()['equation']. Do you have any idea to manipulate the expression with sympy format output? For example, maybe I want to optimize the constant in the expression, or change the operator in the expression.

**Reply by MilesCranmer** [reply]
There are a few other discussion threads on this, perhaps the easiest one to try (hopefully) is https://github.com/MilesCranmer/PySR/discussions/255.

Though it would be really great to have a dedicated feature for this as it has been requested a few times.
